File: ruby/component/dccreceiver.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the same terms as Ruby.

class DccReceiver
  attr_accessor :delegate, :uid, :sender_nick
  attr_accessor :host, :port, :path, :filename, :size, :version
  attr_reader :received_size, :status, :error, :download_filename
  attr_accessor :progress_bar
  
  # RS_WAIT, RS_ERROR, RS_STOP, RS_WAITSTART, RS_WAITCONNECT, 
  # RS_RESUME, RS_RECEIVE, RS_COMPLETE
  
  # waiting, error, stop, connecting
  # receiving, complete
  
  RECORDS_LEN = 10
  
  def initialize
    @size = 0
    @version = 0
    @received_size = 0
    @status = :waiting
    @records = []
    @rec = 0
  end
  
  def speed
    return 0 if @records.empty? || @status != :receiving
    @records.inject(0) {|v,i| v += i }.to_f / @records.length.to_f
  end
  
  def open
    close if @sock
    @records = []
    @rec = 0
    @status = :connecting
    
    @sock = TcpClient.alloc.init
    @sock.delegate = self
    @sock.host = @host
    @sock.port = @port
    @sock.open
  end
  
  def close
    if @sock
      @sock.close
      @sock = nil
    end
    close_file
    @status = :stop if @status != :error && @status != :complete
    @delegate.dccreceiver_on_close(self) if @delegate
  end
  
  
  def tcpclient_on_connect(sender)
    @received_size = 0
    @status = :receiving
    open_file
    @delegate.dccreceiver_on_open(self) if @delegate
    @delegate.dccreceiver_on_change(self) if @delegate
  end
  
  def tcpclient_on_disconnect(sender)
    puts '*** disconnect'
    close
    @delegate.dccreceiver_on_change(self) if @delegate
  end
  
  def tcpclient_on_error(sender, err)
    puts '*** error'
    @status = :error
    @error = err.localizedDescription.to_s
    close_file
    @delegate.dccreceiver_on_close(self) if @delegate
    @delegate.dccreceiver_on_change(self) if @delegate
  end
  
  def tcpclient_on_read(sender)
    s = @sock.read
    @received_size += s.length
    @rec += s.length
    until s.empty? do
      n = @file.write(s)
      s[0...n] = '' if n > 0
    end
    @progress_bar.setDoubleValue(@received_size) if @progress_bar
    if @received_size >= @size
      puts '*** complete ' + @filename
      @status = :complete
      close
      @delegate.dccreceiver_on_change(self) if @delegate
    end
  end
  
  def tcpclient_on_write(sender)
  end
  
  def on_timer
    return if @status != :receiving
    @records << @rec
    @records.shift if @records.length > RECORDS_LEN
    @rec = 0
  end
  
  
  private
  
  PREFIX = '__download__'
  
  def open_file
    return if @file
    base = File.basename(@filename, '.*')
    ext = File.extname(@filename)
    @download_filename = @path + PREFIX + @filename
    i = 0
    while File.exist?(@download_filename)
      @download_filename = @path + PREFIX + base + "_#{i}" + ext
      i += 1
    end
    begin
      @file = File.open(@download_filename, 'w+b')
    rescue
      puts '*** file open error'
      p $!
    end
  end
  
  def close_file
    if @file
      @file.close
      @file = nil
      if @status == :complete
        base = File.basename(@filename, '.*')
        ext = File.extname(@filename)
        fname = @path + @filename
        i = 0
        while File.exist?(fname)
          fname = @path + base + "_#{i}" + ext
          i += 1
        end
        File.rename(@download_filename, fname)
      end
    end
  end
end

File: Classes/Dialogs/DCC/DCCReceiver.m
// LimeChat is copyrighted free software by Satoshi Nakagawa <psychs AT limechat DOT net>.
// You can redistribute it and/or modify it under the terms of the GPL version 2 (see the file GPL.txt).

#import "DCCReceiver.h"


#define DOWNLOADING_PREFIX  @"__download__"
#define RECORDS_LEN         10


@implementation DCCReceiver
{
    TCPClient* _sock;
    NSFileHandle* _file;
    NSMutableArray* _speedRecords;
    double _currentRecord;
}

- (id)init
{
    self = [super init];
    if (self) {
        _speedRecords = [NSMutableArray new];
    }
    return self;
}

- (void)dealloc
{
    [_sock close];
}

- (void)setPath:(NSString *)value
{
    if (_path != value) {
        _path = [value stringByExpandingTildeInPath];
    }
}

- (void)setFileName:(NSString *)value
{
    if (_fileName != value) {
        _fileName = value;
        _icon = [[NSWorkspace sharedWorkspace] iconForFileType:[_fileName pathExtension]];
    }
}

- (double)speed
{
    if (!_speedRecords.count) return 0;

    double sum = 0;
    for (NSNumber* num in _speedRecords) {
        sum += [num doubleValue];
    }
    return sum / _speedRecords.count;
}

- (void)open
{
    if (_sock) {
        [self close];
    }

    _currentRecord = 0;
    [_speedRecords removeAllObjects];

    _sock = [TCPClient new];
    _sock.delegate = self;
    _sock.host = _host;
    _sock.port = _port;
    [_sock open];
}

- (void)close
{
    [_sock close];
    _sock = nil;

    [self closeFile];

    if (_status != DCC_ERROR && _status != DCC_COMPLETE) {
        _status = DCC_STOP;
    }

    [_delegate dccReceiveOnClose:self];
}

- (void)onTimer
{
    if (_status != DCC_RECEIVING) return;

    [_speedRecords addObject:@(_currentRecord)];
    if (_speedRecords.count > RECORDS_LEN) [_speedRecords removeObjectAtIndex:0];
    _currentRecord = 0;
}

- (void)openFile
{
    if (_file) return;

    NSString* base = [_fileName stringByDeletingPathExtension];
    NSString* ext = [_fileName pathExtension];

    NSFileManager* fm = [NSFileManager defaultManager];
    NSString* fullName = [_path stringByAppendingPathComponent:[NSString stringWithFormat:@"%@%@", DOWNLOADING_PREFIX, _fileName]];

    int i = 0;
    while ([fm fileExistsAtPath:fullName]) {
        fullName = [_path stringByAppendingPathComponent:[NSString stringWithFormat:@"%@%@_%d.%@", DOWNLOADING_PREFIX, base, i, ext]];
        ++i;
    }

    NSString* dir = [fullName stringByDeletingLastPathComponent];
    [fm createDirectoryAtPath:dir withIntermediateDirectories:YES attributes:nil error:NULL];
    [fm createFileAtPath:fullName contents:[NSData data] attributes:nil];

    _file = [NSFileHandle fileHandleForUpdatingAtPath:fullName];
    _downloadFileName = fullName;
}

- (void)closeFile
{
    if (!_file) return;

    [_file closeFile];
    _file = nil;

    if (_status == DCC_COMPLETE) {
        NSString* base = [_fileName stringByDeletingPathExtension];
        NSString* ext = [_fileName pathExtension];
        NSString* fullName = [_path stringByAppendingPathComponent:_fileName];

        NSFileManager* fm = [NSFileManager defaultManager];

        int i = 0;
        while ([fm fileExistsAtPath:fullName]) {
            fullName = [_path stringByAppendingPathComponent:[NSString stringWithFormat:@"%@_%d.%@", base, i, ext]];
            ++i;
        }

        [fm moveItemAtPath:_downloadFileName toPath:fullName error:NULL];
        _downloadFileName = fullName;
    }
}

#pragma mark - TCPClient Delegate

- (void)tcpClientDidConnect:(TCPClient*)sender
{
    _processedSize = 0;
    _status = DCC_RECEIVING;

    [self openFile];

    [_delegate dccReceiveOnOpen:self];
}

- (void)tcpClientDidDisconnect:(TCPClient*)sender
{
    if (_status == DCC_COMPLETE || _status == DCC_ERROR) return;

    _status = DCC_ERROR;
    _error = @"Disconnected";
    [self close];

    [_delegate dccReceiveOnError:self];
}

- (void)tcpClient:(TCPClient*)sender error:(NSString*)err
{
    if (_status == DCC_COMPLETE || _status == DCC_ERROR) return;

    _status = DCC_ERROR;
    _error = err;
    [self close];

    [_delegate dccReceiveOnError:self];
}

- (void)tcpClientDidReceiveData:(TCPClient*)sender
{
    NSData* data = [_sock read];
    _processedSize += data.length;
    _currentRecord += data.length;

    if (data.length) {
        [_file writeData:data];
    }

    uint32_t rsize = _processedSize & 0xFFFFFFFF;
    unsigned char ack[4];
    ack[0] = (rsize >> 24) & 0xFF;
    ack[1] = (rsize >> 16) & 0xFF;
    ack[2] = (rsize >>  8) & 0xFF;
    ack[3] = rsize & 0xFF;
    [_sock write:[NSData dataWithBytes:ack length:4]];

    _progressBar.doubleValue = _processedSize;
    [_progressBar setNeedsDisplay:YES];

    if (_processedSize >= _size) {
        _status = DCC_COMPLETE;
        [self close];
        [_delegate dccReceiveOnComplete:self];
    }
}

- (void)tcpClientDidSendData:(TCPClient*)sender
{
}

@end
